src/Microsoft.OData.Core/RawValueWriter.cs

```diff
@@ -154,6 +153,86 @@ internal void Flush()
             }
         }
 
+        /// <summary>
+        /// Asynchronously start writing a raw output. This should only be called once.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        internal async Task StartAsync()
+        {
+            if (this.settings.HasJsonPaddingFunction())
+            {
+                await this.textWriter.WriteAsync(this.settings.JsonPCallback)
+                    .ConfigureAwait(false);
+                await this.textWriter.WriteAsync(JsonConstants.StartPaddingFunctionScope)
+                    .ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously end the writing of a raw output. This should be the last thing called.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        internal Task EndAsync()
+        {
+            if (this.settings.HasJsonPaddingFunction())
+            {
+                return this.textWriter.WriteAsync(JsonConstants.EndPaddingFunctionScope);
+            }
+
+            return TaskUtils.CompletedTask;
+        }
+
+        /// <summary>
+        /// Asynchronously converts the specified <paramref name="value"/> into its raw format and writes it to the output.
+        /// The value has to be of enumeration or primitive type. Only one WriteRawValue call should be made before this object gets disposed.
+        /// </summary>
+        /// <param name="value">The (non-binary) value to write.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        /// <remarks>We do not accept binary values here; WriteBinaryValue should be used for binary data.</remarks>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        internal Task WriteRawValueAsync(object value)
+        {
+            Debug.Assert(!(value is byte[]), "!(value is byte[])");
+
+            string valueAsString;
+            if (value is ODataEnumValue enumValue)
+            {
+                return this.textWriter.WriteAsync(enumValue.Value);
+            }
+            else if (value is Geometry || value is Geography)
+            {
+                return TaskUtils.GetTaskForSynchronousOperation(
+                    () => PrimitiveConverter.Instance.WriteJsonLight(value, jsonWriter));
+            }
+            else if (ODataRawValueUtils.TryConvertPrimitiveToString(value, out valueAsString))
+            {
+                return this.textWriter.WriteAsync(valueAsString);
+            }
+            else
+            {
+                // Value is neither enum nor primitive
+                return TaskUtils.GetFaultedTask(
+                    new ODataException(Strings.ODataUtils_CannotConvertValueToRawString(value.GetType().FullName)));

```

Comment:
Is this better than using `Task.FromException`?

Reply:
`Task.FromException` is not support across all the frameworks that ODL supports